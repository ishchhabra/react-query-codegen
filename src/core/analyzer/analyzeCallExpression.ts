import { NodePath } from "@babel/traverse";
import * as t from "../../utils/babel-types";
import { isReactQueryCall } from "../../utils/react-query";
import { analyzeReferencedIdentifier } from "./analyzeReferencedIdentifier";

export function analyzeCallExpression(
  path: NodePath<t.CallExpression>,
): { statements: t.Statement[]; identifier?: t.Identifier } | undefined | null {
  if (isReactQueryCall(path)) {
    return analyzeReactQueryCall(path);
  }

  return undefined;
}

function analyzeReactQueryCall(
  path: NodePath<t.CallExpression>,
): { statements: t.Statement[]; identifier?: t.Identifier } | undefined | null {
  const statements = [];

  const analysisResult = analyzeReferences(path);
  if (analysisResult === undefined) {
    return undefined;
  }

  if (analysisResult !== null) {
    statements.push(...analysisResult.statements);
  }

  const callExpression = t.callExpression(
    t.memberExpression(t.identifier("queryClient"), t.identifier("fetchQuery")),
    path.node.arguments,
  );

  const parentPath = path.parentPath;
  if (parentPath.isVariableDeclarator()) {
    const identifier = path.scope.generateUidIdentifierBasedOnNode(path.node);
    path.scope.rename((parentPath.node.id as any).name, identifier.name);
    const variableDeclarator = t.variableDeclarator(identifier, callExpression);
    const variableDeclaration = t.variableDeclaration("const", [
      variableDeclarator,
    ]);
    statements.push(variableDeclaration);

    return { statements, identifier };
  } else {
    statements.push(t.expressionStatement(callExpression));
    return { statements };
  }
}

function analyzeReferences(
  path: NodePath,
): { statements: t.Statement[]; identifier?: t.Identifier } | undefined | null {
  const statements = [];

  let errorFound = false;
  path.traverse({
    ReferencedIdentifier(path) {
      const analysisResult = analyzeReferencedIdentifier(path);
      if (analysisResult === undefined) {
        errorFound = true;
      } else if (analysisResult !== null) {
        statements.push(...analysisResult.statements);
      }
    },
  });

  return errorFound ? undefined : { statements };
}
