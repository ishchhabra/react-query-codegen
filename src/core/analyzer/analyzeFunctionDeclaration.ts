import { NodePath } from "@babel/traverse";
import * as t from "../../utils/babel-types";

export function analyzeFunctionDeclaration(
  path: NodePath<t.FunctionDeclaration>
): { statements: t.Statement[] } | undefined | null {
  return { statements: [path.node] };
}
