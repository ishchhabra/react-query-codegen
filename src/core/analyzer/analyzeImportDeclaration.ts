import { NodePath } from "@babel/traverse";
import * as t from "../../utils/babel-types";

export function analyzeImportDeclaration(
  path: NodePath<t.ImportDeclaration>
): { statements: t.Statement[]; identifier?: t.Identifier } | undefined | null {
  return null;
}
