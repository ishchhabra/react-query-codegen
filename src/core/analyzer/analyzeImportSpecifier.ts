import { NodePath } from "@babel/traverse";
import * as t from "../../utils/babel-types";

export function analyzeImportSpecifier(
  path: NodePath<t.ImportSpecifier>,
): { statements: t.Statement[]; identifier?: t.Identifier } | undefined | null {
  return null;
}
