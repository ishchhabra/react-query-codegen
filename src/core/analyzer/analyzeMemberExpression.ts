import { NodePath } from "@babel/traverse";
import * as t from "../../utils/babel-types";

export function analyzeMemberExpression(
  path: NodePath<t.MemberExpression>,
): { statements: t.Statement[] } | undefined | null {
  return { statements: [] };
}
