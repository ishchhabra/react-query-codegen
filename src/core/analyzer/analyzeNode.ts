import { NodePath } from "@babel/traverse";
import * as t from "../../utils/babel-types";
import { analyzeCallExpression } from "./analyzeCallExpression";
import { analyzeFunctionDeclaration } from "./analyzeFunctionDeclaration";
import { analyzeImportDeclaration } from "./analyzeImportDeclaration";
import { analyzeImportSpecifier } from "./analyzeImportSpecifier";
import { analyzeMemberExpression } from "./analyzeMemberExpression";
import { analyzeVariableDeclarator } from "./analyzeVariableDeclarator";

export default function analyzeNode(
  path: NodePath
): { statements: t.Statement[]; identifier?: t.Identifier } | undefined | null {
  switch (true) {
    case path.isCallExpression():
      return analyzeCallExpression(path);
    case path.isFunctionDeclaration():
      return analyzeFunctionDeclaration(path);
    case path.isImportDeclaration():
      return analyzeImportDeclaration(path);
    case path.isImportSpecifier():
      return analyzeImportSpecifier(path);
    case path.isMemberExpression():
      return analyzeMemberExpression(path);
    case path.isVariableDeclarator():
      return analyzeVariableDeclarator(path);
  }

  return undefined;
}
