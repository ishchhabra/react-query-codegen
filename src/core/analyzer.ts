import generate from "@babel/generator";
import { parse } from "@babel/parser";
import * as fs from "fs";
import { FrameworkAdapter } from "../adapters/base";
import * as t from "../utils/babel-types";
import { isReactQueryCall } from "../utils/react-query";
import { analyzeCallExpression } from "./analyzer/analyzeNode";
import { findQueryCalls } from "./analyzer/findQueryCalls";
import {
  DEFAULT_ENCODING,
  DEFAULT_PLUGINS,
  DEFAULT_SOURCE_TYPE,
} from "./constants";
import writeOutput from "./writer";

export function analyzeRoute(route: string, adapter: FrameworkAdapter) {
  const content = fs.readFileSync(route, { encoding: DEFAULT_ENCODING });
  const ast = parse(content, {
    sourceType: DEFAULT_SOURCE_TYPE,
    plugins: DEFAULT_PLUGINS,
  });
  const queryCalls = findQueryCalls(ast);
  console.info(`üîç Found ${queryCalls.length} React Query calls in ${route}`);

  const program: t.Program = t.program([]);

  for (const queryCall of queryCalls) {
    queryCall.traverse({
      MemberExpression(path) {
        const object = path.get("object");
        if (!object.isIdentifier()) {
          return;
        }

        const binding = path.scope.getBinding(object.node.name);
        if (binding === undefined) {
          return;
        }

        if (!binding.path.isVariableDeclarator()) {
          return;
        }

        const init = binding.path.get("init");
        if (!isReactQueryCall(init)) {
          return;
        }

        const property = path.get("property");
        if (!property.isIdentifier()) {
          return;
        }

        if (property.node.name !== "data") {
          queryCalls.splice(queryCalls.indexOf(queryCall), 1);
          return;
        }

        path.replaceWith(object.node);
      },
    });
  }

  for (const queryCall of queryCalls) {
    const analysisResult = analyzeCallExpression(queryCall, program);
    if (analysisResult === undefined) {
      continue;
    }
  }

  const outputPath = adapter.getOutputPath(route, route);
  const output = generate(program);
  writeOutput(route, outputPath, adapter, output.code);
}
