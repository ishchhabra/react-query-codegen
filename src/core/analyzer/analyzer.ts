import generate from "@babel/generator";
import { parse } from "@babel/parser";
import fs from "fs";
import { FrameworkAdapter } from "../../adapters/base";
import * as t from "../../utils/babel-types";
import {
  DEFAULT_ENCODING,
  DEFAULT_PLUGINS,
  DEFAULT_SOURCE_TYPE,
} from "../constants";
import writeOutput from "../writer";
import analyzeNode from "./analyzeNode";
import { findQueryCalls } from "./findQueryCalls";

export function analyzeRoute(route: string, adapter: FrameworkAdapter) {
  const content = fs.readFileSync(route, DEFAULT_ENCODING);
  const ast = parse(content, {
    sourceType: DEFAULT_SOURCE_TYPE,
    plugins: DEFAULT_PLUGINS,
  });
  const queryCalls = findQueryCalls(ast);
  console.info(`üîç Found ${queryCalls.length} React Query calls in ${route}`);

  const statements: t.Statement[] = [];
  for (const queryCall of queryCalls) {
    const analysisResult = analyzeNode(queryCall);
    if (analysisResult === undefined || analysisResult === null) {
      continue;
    }

    statements.push(...analysisResult.statements);
  }

  const queryClientIdentifier = t.identifier("queryClient");
  queryClientIdentifier.typeAnnotation = t.typeAnnotation(
    t.genericTypeAnnotation(t.identifier("QueryClient"), null)
  );
  const prefetchQueryFunctionDeclaration = t.functionDeclaration(
    t.identifier("prefetchQuery"),
    [queryClientIdentifier],
    t.blockStatement(statements)
  );
  const exportNamedDeclaration = t.exportNamedDeclaration(
    prefetchQueryFunctionDeclaration
  );
  const importDeclaration = t.importDeclaration(
    [
      t.importSpecifier(
        t.identifier("QueryClient"),
        t.identifier("QueryClient")
      ),
    ],
    t.stringLiteral("@tanstack/react-query")
  );
  const program = t.program([importDeclaration, exportNamedDeclaration]);

  const outputPath = adapter.getOutputPath(route, route);
  const outputContent = generate(program).code;
  writeOutput(route, outputPath, adapter, outputContent);
}
