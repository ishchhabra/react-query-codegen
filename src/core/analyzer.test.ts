import path from "path";

import fs from "fs";
import { NextAppRouterAdapter } from "../adapters/next/app-router";
import { analyzeRoute } from "./analyzer/analyzer";

describe("analyzeRoute", () => {
  let adapter: NextAppRouterAdapter;

  beforeEach(() => {
    adapter = new NextAppRouterAdapter();
    jest.clearAllMocks();
  });

  it("should analyze a basic query call properly", () => {
    const writeFileSyncMock = jest
      .spyOn(fs, "writeFileSync")
      .mockImplementation();

    analyzeRoute(
      path.join(__dirname, "__fixtures__/next-app/app/basic-call/page.tsx"),
      adapter
    );

    const [_, outputContent] = writeFileSyncMock.mock.calls[0];
    expect(outputContent).toMatchSnapshot();
  });

  it("should analyze a basic query call with variable properly", () => {
    const writeFileSyncMock = jest
      .spyOn(fs, "writeFileSync")
      .mockImplementation();

    analyzeRoute(
      path.join(
        __dirname,
        "__fixtures__/next-app/app/basic-call-with-variable/page.tsx"
      ),
      adapter
    );

    const [_, outputContent] = writeFileSyncMock.mock.calls[0];
    expect(outputContent).toMatchSnapshot();
  });

  it("should analyze a query dependent on another query properly", () => {
    const writeFileSyncMock = jest
      .spyOn(fs, "writeFileSync")
      .mockImplementation();

    analyzeRoute(
      path.join(
        __dirname,
        "__fixtures__/next-app/app/query-dependent-on-query/page.tsx"
      ),
      adapter
    );

    const [_, outputContent] = writeFileSyncMock.mock.calls[0];
    expect(outputContent).toMatchSnapshot();
  });

  it("should analyze a query dependent on useState properly", () => {
    const writeFileSyncMock = jest
      .spyOn(fs, "writeFileSync")
      .mockImplementation();

    analyzeRoute(
      path.join(
        __dirname,
        "__fixtures__/next-app/app/query-dependent-on-use-state/page.tsx"
      ),
      adapter
    );

    const [_, outputContent] = writeFileSyncMock.mock.calls[0];
    expect(outputContent).toMatchSnapshot();
  });
});
