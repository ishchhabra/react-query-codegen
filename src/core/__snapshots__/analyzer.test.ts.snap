// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`analyzeRoute should analyze a basic query call properly 1`] = `
"import { QueryClient } from "@tanstack/react-query";
export function prefetchQuery(queryClient: QueryClient) {
  queryClient.fetchQuery({
    queryKey: ["todos"],
    queryFn: () => ["3"]
  });
}"
`;

exports[`analyzeRoute should analyze a basic query call with variable properly 1`] = `
"import { QueryClient } from "@tanstack/react-query";
export function prefetchQuery(queryClient: QueryClient) {
  const _temp = "5";
  queryClient.fetchQuery({
    queryKey: ["todos"],
    queryFn: () => _temp
  });
}"
`;

exports[`analyzeRoute should analyze a query dependent on another query properly 1`] = `
"import { QueryClient } from "@tanstack/react-query";
export function prefetchQuery(queryClient: QueryClient) {
  const _useQuery = queryClient.fetchQuery({
    queryKey: ["todos"],
    queryFn: () => "result1"
  });
  queryClient.fetchQuery({
    queryKey: ["todos"],
    queryFn: () => _useQuery + "result2"
  });
}"
`;

exports[`analyzeRoute should analyze a query dependent on useState properly 1`] = `
"import { QueryClient } from "@tanstack/react-query";
export function prefetchQuery(queryClient: QueryClient) {
  const _useState = 0;
  queryClient.fetchQuery({
    queryKey: ["todos"],
    queryFn: () => _useState
  });
}"
`;
